File: sto-agents-complete/services/teams_notify.py
"""
Teams Notification Service
Sends notifications to Microsoft Teams via webhooks.

Setup:
1. Create an Incoming Webhook in your Teams channel
2. Set TEAMS_WEBHOOK_URL environment variable
   OR update config/settings.py with the webhook URL
"""

import logging
import json
import os
from datetime import datetime
from typing import Optional
import urllib.request
import urllib.error

# Setup logging
logger = logging.getLogger('teams_notify')

# Import settings
try:
    from config.settings import TEAMS_WEBHOOK_URL, TEAMS_WEBHOOKS
except ImportError:
    TEAMS_WEBHOOK_URL = os.environ.get("TEAMS_WEBHOOK_URL", "")
    TEAMS_WEBHOOKS = {}

# =============================================================================
# WEBHOOK CONFIGURATION
# =============================================================================

def get_webhook_url(channel: str = "general") -> str:
    """Get webhook URL for a specific channel"""
    # Check channel-specific webhooks first
    if TEAMS_WEBHOOKS and channel in TEAMS_WEBHOOKS:
        url = TEAMS_WEBHOOKS[channel]
        if url:
            return url
    
    # Fall back to default webhook
    if TEAMS_WEBHOOK_URL:
        return TEAMS_WEBHOOK_URL
    
    # Check environment variable
    return os.environ.get("TEAMS_WEBHOOK_URL", "")

# =============================================================================
# MESSAGE FORMATTING
# =============================================================================

def format_adaptive_card(title: str, message: str, color: str = "0076D7") -> dict:
    """
    Format message as Teams Adaptive Card.
    
    Args:
        title: Card title
        message: Main message content (supports markdown)
        color: Accent color (hex without #)
    """
    return {
        "@type": "MessageCard",
        "@context": "http://schema.org/extensions",
        "themeColor": color,
        "summary": title,
        "sections": [{
            "activityTitle": title,
            "activitySubtitle": datetime.now().strftime("%B %d, %Y %I:%M %p"),
            "text": message,
            "markdown": True
        }]
    }

def format_simple_message(message: str) -> dict:
    """Format a simple text message"""
    return {
        "text": message
    }

# =============================================================================
# SEND FUNCTIONS
# =============================================================================

def send_teams_message(payload: dict, channel: str = "general") -> bool:
    """
    Send a message payload to Teams.
    
    Args:
        payload: Message payload (dict)
        channel: Target channel (general, alerts, automation)
    
    Returns:
        True if successful, False otherwise
    """
    webhook_url = get_webhook_url(channel)
    
    if not webhook_url:
        logger.warning("Teams webhook URL not configured - skipping notification")
        return False
    
    try:
        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            webhook_url,
            data=data,
            headers={'Content-Type': 'application/json'}
        )
        
        with urllib.request.urlopen(req, timeout=10) as response:
            if response.status == 200:
                logger.info(f"Teams notification sent to {channel}")
                return True
            else:
                logger.error(f"Teams webhook returned status {response.status}")
                return False
                
    except urllib.error.URLError as e:
        logger.error(f"Failed to send Teams notification: {e}")
        return False
    except Exception as e:
        logger.error(f"Unexpected error sending Teams notification: {e}")
        return False

def send_teams_notification(title: str, message: str, channel: str = "general", 
                           color: str = "0076D7") -> bool:
    """
    Send a formatted notification to Teams.
    
    Args:
        title: Notification title
        message: Message content (supports basic markdown)
        channel: Target channel
        color: Accent color (hex without #)
            - 0076D7: Blue (default/info)
            - 00FF00: Green (success)
            - FFA500: Orange (warning)
            - FF0000: Red (error)
    """
    payload = format_adaptive_card(title, message, color)
    return send_teams_message(payload, channel)

# =============================================================================
# SPECIFIC NOTIFICATION TYPES
# =============================================================================

def notify_plan_intake(plans_downloaded: list, errors: list = None) -> bool:
    """Notify about plan intake results"""
    if not plans_downloaded and not errors:
        return True  # Nothing to report
    
    message_parts = []
    
    if plans_downloaded:
        message_parts.append(f"**{len(plans_downloaded)} plans downloaded:**")
        for plan in plans_downloaded[:10]:  # Limit to 10
            message_parts.append(f"• {plan}")
        if len(plans_downloaded) > 10:
            message_parts.append(f"• ... and {len(plans_downloaded) - 10} more")
    
    if errors:
        message_parts.append(f"\n⚠️ **{len(errors)} errors:**")
        for error in errors[:5]:
            message_parts.append(f"• {error}")
    
    color = "00FF00" if not errors else "FFA500"
    return send_teams_notification("Plan Intake Complete", "\n".join(message_parts), color=color)

def notify_completeness_check(total: int, complete: int, incomplete: int, 
                             incomplete_jobs: list = None) -> bool:
    """Notify about completeness check results"""
    message_parts = [
        f"**Summary:**",
        f"• Total Jobs: {total}",
        f"• Complete: {complete} ✓",
        f"• Incomplete: {incomplete} ⚠️",
    ]
    
    if incomplete_jobs and incomplete > 0:
        message_parts.append(f"\n**Incomplete Jobs:**")
        for job in incomplete_jobs[:10]:
            message_parts.append(f"• {job.get('subdivision', '?')} - {job.get('lot', '?')}")
            if job.get('missing'):
                message_parts.append(f"  Missing: {', '.join(job['missing'][:3])}")
    
    color = "00FF00" if incomplete == 0 else "FFA500"
    return send_teams_notification("Completeness Check", "\n".join(message_parts), color=color)

def notify_pdss_updated(new_jobs: int = 0, updated_jobs: int = 0, 
                       archived_jobs: int = 0) -> bool:
    """Notify about PDSS sync results"""
    message_parts = [
        "**PDSS Sync Complete**",
        f"• New Jobs: {new_jobs}",
        f"• Updated: {updated_jobs}",
        f"• Archived: {archived_jobs}",
    ]
    
    color = "0076D7"  # Blue for info
    return send_teams_notification("PDSS Sync", "\n".join(message_parts), color=color)

def notify_backup_complete(backup_type: str, files: int, size_mb: float, 
                          destination: str = None) -> bool:
    """Notify about backup completion"""
    message_parts = [
        f"**{backup_type.title()} Backup Complete**",
        f"• Files: {files:,}",
        f"• Size: {size_mb:.1f} MB",
    ]
    
    if destination:
        message_parts.append(f"• Location: {destination}")
    
    return send_teams_notification("Backup Complete", "\n".join(message_parts), color="00FF00")

def notify_error(agent_name: str, error_message: str, details: str = None) -> bool:
    """Send error notification"""
    message_parts = [
        f"**Agent:** {agent_name}",
        f"**Error:** {error_message}",
    ]
    
    if details:
        message_parts.append(f"\n**Details:**\n{details[:500]}")
    
    return send_teams_notification(
        f"⚠️ Agent Error: {agent_name}", 
        "\n".join(message_parts), 
        channel="alerts",
        color="FF0000"
    )

def notify_supplypro_report(dashboard: dict, epo_summary: dict, 
                           documents: dict = None) -> bool:
    """Send SupplyPro daily report notification"""
    message_parts = [
        "**Dashboard:**",
        f"• New Orders: {dashboard.get('new_orders', 0):,}",
        f"• To Do: {dashboard.get('to_do', 0):,}",
        f"• Change Orders: {dashboard.get('change_orders', 0)}",
        "",
        "**EPO Summary:**",
        f"• Total: {epo_summary.get('total', 0)} (${epo_summary.get('total_amount', 0):,.2f})",
        f"• Pending: {epo_summary.get('pending', 0)} (${epo_summary.get('pending_amount', 0):,.2f})",
    ]
    
    if documents:
        message_parts.append(f"\n**New Documents:** {documents.get('total_unviewed', 0)} unviewed")
    
    color = "FFA500" if epo_summary.get('pending', 0) > 0 else "0076D7"
    return send_teams_notification("SupplyPro Daily Report", "\n".join(message_parts), color=color)

# =============================================================================
# TESTING
# =============================================================================

def test_webhook() -> bool:
    """Test the webhook connection"""
    return send_teams_notification(
        "🔧 STO Agents Test",
        "This is a test notification from STO Agents.\n\nIf you can see this, your webhook is configured correctly!",
        color="00FF00"
    )

# =============================================================================
# CLI
# =============================================================================

if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Teams Notification Service")
    parser.add_argument("--test", action="store_true", help="Send test notification")
    parser.add_argument("--message", type=str, help="Send custom message")
    parser.add_argument("--title", type=str, default="STO Agents", help="Message title")
    
    args = parser.parse_args()
    
    if args.test:
        success = test_webhook()
        print(f"Test notification {'sent successfully' if success else 'FAILED'}")
    elif args.message:
        success = send_teams_notification(args.title, args.message)
        print(f"Notification {'sent successfully' if success else 'FAILED'}")
    else:
        webhook_url = get_webhook_url()
        if webhook_url:
            print(f"Webhook configured: {webhook_url[:50]}...")
        else:
            print("⚠️ No webhook URL configured!")
            print("\nTo configure:")
            print("1. Set TEAMS_WEBHOOK_URL environment variable")
            print("2. Or update config/settings.py")

File: backend/src/services/teamsNotify.ts
/**
 * Teams Notification Service
 *
 * Sends notifications to Microsoft Teams via webhooks.
 * Mirrors the Python teams_notify.py functionality for web-based notifications.
 *
 * Setup:
 * 1. Create an Incoming Webhook in your Teams channel
 * 2. Set TEAMS_WEBHOOK_URL environment variable
 *
 * Channels:
 * - general: Default channel for regular updates
 * - alerts: Critical alerts and errors
 * - automation: Automation status updates
 */

import https from 'https';
import http from 'http';

// =============================================================================
// TYPES
// =============================================================================

interface TeamsSection {
  activityTitle: string;
  activitySubtitle?: string;
  text: string;
  markdown: boolean;
  facts?: { name: string; value: string }[];
}

interface TeamsMessageCard {
  '@type': 'MessageCard';
  '@context': string;
  themeColor: string;
  summary: string;
  sections: TeamsSection[];
}

interface SimpleMessage {
  text: string;
}

type TeamsPayload = TeamsMessageCard | SimpleMessage;

export interface NotificationResult {
  success: boolean;
  channel: string;
  error?: string;
}

// Color constants
export const NotificationColors = {
  INFO: '0076D7',     // Blue
  SUCCESS: '00FF00',  // Green
  WARNING: 'FFA500',  // Orange
  ERROR: 'FF0000',    // Red
} as const;

type NotificationColor = typeof NotificationColors[keyof typeof NotificationColors];

// =============================================================================
// WEBHOOK CONFIGURATION
// =============================================================================

function getWebhookUrl(channel: string = 'general'): string | null {
  // Check channel-specific webhooks
  const channelEnvKey = `TEAMS_WEBHOOK_URL_${channel.toUpperCase()}`;
  const channelUrl = process.env[channelEnvKey];
  if (channelUrl) {
    return channelUrl;
  }

  // Fall back to default webhook
  const defaultUrl = process.env.TEAMS_WEBHOOK_URL;
  if (defaultUrl) {
    return defaultUrl;
  }

  return null;
}

// =============================================================================
// MESSAGE FORMATTING
// =============================================================================

function formatAdaptiveCard(
  title: string,
  message: string,
  color: string = NotificationColors.INFO
): TeamsMessageCard {
  const now = new Date();
  const subtitle = now.toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
  });

  return {
    '@type': 'MessageCard',
    '@context': 'http://schema.org/extensions',
    themeColor: color,
    summary: title,
    sections: [
      {
        activityTitle: title,
        activitySubtitle: subtitle,
        text: message,
        markdown: true,
      },
    ],
  };
}

function formatSimpleMessage(message: string): SimpleMessage {
  return { text: message };
}

// =============================================================================
// SEND FUNCTIONS
// =============================================================================

async function sendTeamsMessage(
  payload: TeamsPayload,
  channel: string = 'general'
): Promise<NotificationResult> {
  const webhookUrl = getWebhookUrl(channel);

  if (!webhookUrl) {
    console.log(`[TeamsNotify] Webhook not configured for channel: ${channel} - skipping notification`);
    return {
      success: false,
      channel,
      error: 'Webhook URL not configured',
    };
  }

  return new Promise((resolve) => {
    try {
      const url = new URL(webhookUrl);
      const data = JSON.stringify(payload);

      const options: https.RequestOptions = {
        hostname: url.hostname,
        port: url.port || (url.protocol === 'https:' ? 443 : 80),
        path: url.pathname + url.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(data),
        },
        timeout: 10000,
      };

      const protocol = url.protocol === 'https:' ? https : http;

      const req = protocol.request(options, (res) => {
        if (res.statusCode === 200) {
          console.log(`[TeamsNotify] Notification sent to ${channel}`);
          resolve({ success: true, channel });
        } else {
          console.error(`[TeamsNotify] Webhook returned status ${res.statusCode}`);
          resolve({
            success: false,
            channel,
            error: `HTTP ${res.statusCode}`,
          });
        }
      });

      req.on('error', (error) => {
        console.error(`[TeamsNotify] Failed to send notification: ${error.message}`);
        resolve({
          success: false,
          channel,
          error: error.message,
        });
      });

      req.on('timeout', () => {
        req.destroy();
        console.error('[TeamsNotify] Request timeout');
        resolve({
          success: false,
          channel,
          error: 'Request timeout',
        });
      });

      req.write(data);
      req.end();
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error';
      console.error(`[TeamsNotify] Unexpected error: ${errorMessage}`);
      resolve({
        success: false,
        channel,
        error: errorMessage,
      });
    }
  });
}

/**
 * Send a formatted notification to Teams
 */
export async function sendTeamsNotification(
  title: string,
  message: string,
  channel: string = 'general',
  color: NotificationColor = NotificationColors.INFO
): Promise<NotificationResult> {
  const payload = formatAdaptiveCard(title, message, color);
  return sendTeamsMessage(payload, channel);
}

// =============================================================================
// SPECIFIC NOTIFICATION TYPES
// =============================================================================

/**
 * Notify about plan intake results
 */
export async function notifyPlanIntake(
  plansDownloaded: string[],
  errors: string[] = []
): Promise<NotificationResult> {
  if (!plansDownloaded.length && !errors.length) {
    return { success: true, channel: 'general' };
  }

  const messageParts: string[] = [];

  if (plansDownloaded.length) {
    messageParts.push(`**${plansDownloaded.length} plans downloaded:**`);
    const displayPlans = plansDownloaded.slice(0, 10);
    displayPlans.forEach((plan) => messageParts.push(`- ${plan}`));
    if (plansDownloaded.length > 10) {
      messageParts.push(`- ... and ${plansDownloaded.length - 10} more`);
    }
  }

  if (errors.length) {
    messageParts.push(`\n**${errors.length} errors:**`);
    const displayErrors = errors.slice(0, 5);
    displayErrors.forEach((error) => messageParts.push(`- ${error}`));
  }

  const color = errors.length ? NotificationColors.WARNING : NotificationColors.SUCCESS;
  return sendTeamsNotification('Plan Intake Complete', messageParts.join('\n'), 'general', color);
}

/**
 * Notify about completeness check results
 */
export async function notifyCompletenessCheck(
  total: number,
  complete: number,
  incomplete: number,
  incompleteJobs?: { subdivision: string; lot: string; missing?: string[] }[]
): Promise<NotificationResult> {
  const messageParts: string[] = [
    '**Summary:**',
    `- Total Jobs: ${total}`,
    `- Complete: ${complete}`,
    `- Incomplete: ${incomplete}`,
  ];

  if (incompleteJobs && incomplete > 0) {
    messageParts.push('\n**Incomplete Jobs:**');
    const displayJobs = incompleteJobs.slice(0, 10);
    displayJobs.forEach((job) => {
      messageParts.push(`- ${job.subdivision} - ${job.lot}`);
      if (job.missing?.length) {
        messageParts.push(`  Missing: ${job.missing.slice(0, 3).join(', ')}`);
      }
    });
  }

  const color = incomplete === 0 ? NotificationColors.SUCCESS : NotificationColors.WARNING;
  return sendTeamsNotification('Completeness Check', messageParts.join('\n'), 'general', color);
}

/**
 * Notify about PDSS sync results
 */
export async function notifyPDSSUpdated(
  newJobs: number = 0,
  updatedJobs: number = 0,
  archivedJobs: number = 0
): Promise<NotificationResult> {
  const messageParts: string[] = [
    '**PDSS Sync Complete**',
    `- New Jobs: ${newJobs}`,
    `- Updated: ${updatedJobs}`,
    `- Archived: ${archivedJobs}`,
  ];

  return sendTeamsNotification('PDSS Sync', messageParts.join('\n'), 'general', NotificationColors.INFO);
}

/**
 * Notify about job status changes
 */
export async function notifyJobStatusChange(
  jobId: string,
  subdivision: string,
  lot: string,
  oldStatus: string,
  newStatus: string,
  changedBy?: string
): Promise<NotificationResult> {
  const messageParts: string[] = [
    `**Job:** ${subdivision} - ${lot}`,
    `**Status Change:** ${oldStatus} -> ${newStatus}`,
  ];

  if (changedBy) {
    messageParts.push(`**Changed By:** ${changedBy}`);
  }

  return sendTeamsNotification('Job Status Updated', messageParts.join('\n'), 'general', NotificationColors.INFO);
}

/**
 * Notify about backup completion
 */
export async function notifyBackupComplete(
  backupType: string,
  files: number,
  sizeMb: number,
  destination?: string
): Promise<NotificationResult> {
  const messageParts: string[] = [
    `**${backupType.charAt(0).toUpperCase() + backupType.slice(1)} Backup Complete**`,
    `- Files: ${files.toLocaleString()}`,
    `- Size: ${sizeMb.toFixed(1)} MB`,
  ];

  if (destination) {
    messageParts.push(`- Location: ${destination}`);
  }

  return sendTeamsNotification('Backup Complete', messageParts.join('\n'), 'general', NotificationColors.SUCCESS);
}

/**
 * Send error notification
 */
export async function notifyError(
  agentName: string,
  errorMessage: string,
  details?: string
): Promise<NotificationResult> {
  const messageParts: string[] = [
    `**Agent:** ${agentName}`,
    `**Error:** ${errorMessage}`,
  ];

  if (details) {
    messageParts.push(`\n**Details:**\n${details.substring(0, 500)}`);
  }

  return sendTeamsNotification(
    `Agent Error: ${agentName}`,
    messageParts.join('\n'),
    'alerts',
    NotificationColors.ERROR
  );
}

/**
 * Notify about SupplyPro daily report
 */
export async function notifySupplyProReport(
  dashboard: { newOrders?: number; toDo?: number; changeOrders?: number },
  epoSummary: { total?: number; totalAmount?: number; pending?: number; pendingAmount?: number },
  documents?: { totalUnviewed?: number }
): Promise<NotificationResult> {
  const messageParts: string[] = [
    '**Dashboard:**',
    `- New Orders: ${(dashboard.newOrders ?? 0).toLocaleString()}`,
    `- To Do: ${(dashboard.toDo ?? 0).toLocaleString()}`,
    `- Change Orders: ${dashboard.changeOrders ?? 0}`,
    '',
    '**EPO Summary:**',
    `- Total: ${epoSummary.total ?? 0} ($${(epoSummary.totalAmount ?? 0).toLocaleString(undefined, { minimumFractionDigits: 2 })})`,
    `- Pending: ${epoSummary.pending ?? 0} ($${(epoSummary.pendingAmount ?? 0).toLocaleString(undefined, { minimumFractionDigits: 2 })})`,
  ];

  if (documents?.totalUnviewed) {
    messageParts.push(`\n**New Documents:** ${documents.totalUnviewed} unviewed`);
  }

  const color = (epoSummary.pending ?? 0) > 0 ? NotificationColors.WARNING : NotificationColors.INFO;
  return sendTeamsNotification('SupplyPro Daily Report', messageParts.join('\n'), 'general', color);
}

/**
 * Test the webhook connection
 */
export async function testWebhook(): Promise<NotificationResult> {
  return sendTeamsNotification(
    'MindFlow Test',
    'This is a test notification from MindFlow.\n\nIf you can see this, your webhook is configured correctly!',
    'general',
    NotificationColors.SUCCESS
  );
}

// =============================================================================
// SERVICE CLASS (alternative OOP interface)
// =============================================================================

class TeamsNotifyService {
  async send(title: string, message: string, channel?: string, color?: NotificationColor) {
    return sendTeamsNotification(title, message, channel, color);
  }

  async notifyPlanIntake(plans: string[], errors?: string[]) {
    return notifyPlanIntake(plans, errors);
  }

  async notifyPDSSUpdated(newJobs: number, updatedJobs: number, archivedJobs: number) {
    return notifyPDSSUpdated(newJobs, updatedJobs, archivedJobs);
  }

  async notifyJobStatusChange(
    jobId: string,
    subdivision: string,
    lot: string,
    oldStatus: string,
    newStatus: string,
    changedBy?: string
  ) {
    return notifyJobStatusChange(jobId, subdivision, lot, oldStatus, newStatus, changedBy);
  }

  async notifyError(agentName: string, errorMessage: string, details?: string) {
    return notifyError(agentName, errorMessage, details);
  }

  async test() {
    return testWebhook();
  }

  isConfigured(channel: string = 'general'): boolean {
    return getWebhookUrl(channel) !== null;
  }
}

// Export singleton instance
export const teamsNotifyService = new TeamsNotifyService();
export default TeamsNotifyService;
